### ISSUE TITLE
OpenGL Deferred Shading issue 

### ISSUE THREAD

**Issue by stephane-lb** [inviwo/inviwo#13]
Hi,

I just updated inviwo on my mac and I really like the style of your apps, it would be great if more and more people could start to use it. 

Just a small bug that I just notice : 
<img width="518" alt="screen shot 2017-08-01 at 16 50 57" src="https://user-images.githubusercontent.com/5918598/28834362-f2d98fd8-76d9-11e7-9650-f0ca3f90b243.png">

And I will add the trace provide in the console :
16:50:03.770	inviwo.cpp	1 errors generated during startup
16:50:20.818	OpenGL	GL_INVALID_ENUM: An unacceptable value is specified for an enumerated argument. The offending command is ignored and has no other side effect than to set the error flag.
16:50:20.822	Shader	Id: 4 
WARNING: Could not find vertex shader attribute 'in_Normal' to match BindAttributeLocation request.
WARNING: Output of vertex shader 'color_' not read by fragment shader
WARNING: Output of vertex shader 'texCoord_' not read by fragment shader
16:50:20.826	Shader	Id: 7 
WARNING: Could not find vertex shader attribute 'in_Normal' to match BindAttributeLocation request.
WARNING: Output of vertex shader 'texCoord_' not read by fragment shader
16:50:20.845	OpenGL Debug	Debugging off for context: Canvas (0x7fba7e24f460)
16:50:20.868	Shader	Id: 10 
WARNING: Could not find vertex shader attribute 'in_Color' to match BindAttributeLocation request.
WARNING: Could not find vertex shader attribute 'in_Normal' to match BindAttributeLocation request.
WARNING: Could not find fragment shader output 'PickingData' to match FragDataBinding request.

Hope that will help. 

Best Regards,

stephane-lb

**Comment by petersteneteg** [comment]
Thanks!

To get the filename a line etc in the log you can right-click in the console and select columns to show.
You can also run inviwo from the command line with
```
   -c,  --logconsole
     Write log messages to console (cout)

   -l <logfile>,  --logfile <logfile>
     Write log messages to file.
```
On you mac you will have to run ```inviwo.app/Contents/MacOS/inviwo```

Most of the shader warnings should be fine, the mac drivers seems to like to print a lot of shader warnings.

The main issue seems to be with the OpenCL Driver "CL_INVALID_DEVICE" etc. 
Can you share what physical setup you have? We have had some issues with the intel devices that are built into may mac before might be related? You can select which OpenCL device to use in the OpenCL Settings dialog in the setting pane  (View->Settings), you might test if changing device there makes any difference.

**Comment by petersteneteg** [comment]
@ResearchDaniel do you have any insight?

**Comment by stephane-lb** [comment]
Hi Peter,

Thanks for your answer, and really sorry did not think about trying a -h.

I am using a Macbook pro from early 2013 ( As seen in the picture )

<img width="1440" alt="screen shot 2017-07-27 at 19 27 30" src="https://user-images.githubusercontent.com/5918598/28884047-5bff1860-77a8-11e7-84d9-7706d2bc0e54.png">

I tried changing to NVIDIA GT 750 M for openCL but nothing changed. I am checking that's I installed it properly.

The issue is appear each time you load a scene in the canvaGL with the first vertex shader during the binding. It could be this or an issue with the context ... I am trying to understand the pipeline.

Best Regards,

Stéphane Le Boeuf

**Comment by petersteneteg** [comment]
That is strange, I think we have tested it on similar hardware successfully. As a work around you could try and disable all OpenCL related modules in CMake, (IVW_MODULE_OPENCL, IVW_MODULE_BASECL) and see if everything else works.
You might run it with the -l filename option and upload the complete log here for us if possible.

**Comment by stephane-lb** [comment]
Hi Peter,

Thanks for your answer.

About the logfile sadly none of these commands generated the log file
``` bash
./inviwo -l inviwo.log 
#or
./inviwo -l "./inviwo.log"
```

By disabling OpenCL, the error about OpenCL disappeared, however the issue in the  canvas is the same.

**Comment by stephane-lb** [comment]
Hum, I just added a breakpoint in canvaGL.cpp:130 and the getCanvasDimenions().x and getCanvasDimenions().y are half the size.

Just made a quick fix :    
```cpp
glViewport(0, 0, static_cast<GLsizei>(2*getCanvasDimensions().x),
               static_cast<GLsizei>(2*getCanvasDimensions().y));
```
( Fix heighfield example and picking example )

It looks like the same issue than the Qt interface which looks really big due to the problem with Retina screen. Waiting to hear from you, to continue to investigate.

**Comment by petersteneteg** [comment]
I see that is interesting, I guess you are running with a retina display. We do most of out OSX testing on  a mac pro, so we are not able to do much retina testing. 

From retained a bit of qt docs i suspect that the factor 2 above is the devicePixelRatio (http://doc.qt.io/qt-5/qwindow.html#devicePixelRatio). Unfortunately that is not available in CanvasGL, but it you go into CanvasQt in canvasqt.h inviwo_module_openglqt. The structure is a bit complicated, since we played with different qt backends for OpenGL.

We have a template class CanvasQtBase that is templated on a qt backend using CRTP, right now CanvasQGLWidget, which derives from QGLWidget and CanvasGL, 
so CanvasQt = CanvasQtBase<CanvasQGLWidget>;

To get a device pixel ratio into I suppose we have to add a virtual method into CanvasGL and implement that in CanvasQt. You are most welcome to try this out, since I at the moment have very limited testing possibilities.

As you mentioned before I guess we might need to update some plist files too

**Comment by stephane-lb** [comment]
Hey Peter,

Yes, lucky me, I have a Retina, I will continue to look for it and let you know. If I understand your code it is also possible to switch for a GLFW viewer ( in the CMakeList ), so I tried this and the issue is the same. 

I think this is a clue of what is happening, I will try to find it and publish a fix.

Have a good day.

**Comment by petersteneteg** [comment]
Yes, you can also run the GLFW_minimum app which is just a viewer, but runs with out any qt. But the issue should be the same.  
According to glfw you can work with the frame buffer size instead of the window size 
http://www.glfw.org/docs/latest/window.html#window_fbsize

**Comment by stephane-lb** [comment]
Hi Peter,

A quick fix was to : 
-> make the getCanvasDimensions *virtual* in canvas
-> *override* it in CanvasQGLWidget  with something like this : 

```cpp
size2_t CanvasQGLWidget::getCanvasDimensions() const
{
  size2_t canvasDimension = Canvas::getCanvasDimensions();
  int pixelRatio = this->devicePixelRatio();
  return size2_t(canvasDimension.x * pixelRatio, canvasDimension.y * pixelRatio);
}
```

Would that seems clean enough for you ?

**Comment by petersteneteg** [comment]
I susspect that, as you noted, that the dimensions are use in other places to, and that migth create isuees. Would have to investigate how the corrdinates of mouse events from qt are handled, in virtual pixels or device pixels?

**Comment by stephane-lb** [comment]
Any objection to create 2 methods : 
- GetCanvasDimensions() << As usual
- GetViewportDimensions() << Using for the GLViewport.

That should solve all the issues.

**Comment by stephane-lb** [comment]
Sorry for not replying, Qt handle mouse coordinate in Pixel coordinates. Which explain why the mix.

**Comment by scanberg** [comment]
The fix you have suggested plus this additional fix seems to solve the issues for now:

In canvasqt.h, change normalPos to the following:

`template <typename T>
dvec2 inviwo::CanvasQtBase<T>::normalPos(dvec2 pos) const {
    auto scale = this->devicePixelRatio();
    return util::invertY(pos * dvec2(scale), this->getCanvasDimensions()) /
        dvec2(this->getCanvasDimensions() - size2_t(1));
}`

Tis could work as a fix for now, but needs additional testing before its integrated...

**Comment by stephane-lb** [comment]
( Sorry to have been quiet testing that today )

**Comment by stephane-lb** [comment]
@scanberg was this pushed ? Just pull the new version of the code and everything works fine on my mac. Only one new issue with the trackpad, it seems that;s it is detected as a "click".

Thank you very much.
